File: FutureVoice/Views/DialogueLine.swift
import SwiftUI

/// Who is speaking a dialogue line. Top-level (not nested in `DialogueLine`)
/// so callers can read `alignment` without spelling out generic parameters.
enum DialogueSpeaker {
    /// The learner. Trailing side, accent fill.
    case user
    /// The fluent self, or a Watch counterpart. Leading side, neutral fill.
    case other

    var isUser: Bool { self == .user }

    /// Which side this speaker's content sits on. Exposed so a surface's own
    /// accessory stack (Meaning / Shadow / suggestion chips) lines up with the
    /// bubble instead of re-deriving the rule.
    var alignment: HorizontalAlignment { isUser ? .trailing : .leading }
}

/// The ONE dialogue-line surface, shared by every conversation view — Watch's
/// scripted dialogue, the live call transcript, and past-conversation detail.
///
/// Speaker separation (the name label, which side the line sits on, the bubble
/// fill, the playback ring) is defined HERE and nowhere else, so restyling the
/// conversation UI later is a single-file change. Callers supply only their own
/// line content (plain text, attributed text with tappable words, …) and their
/// per-surface accessories — never the chrome.
struct DialogueLine<Content: View, Accessory: View>: View {

    /// Text size. The call screen is read at arm's length mid-conversation;
    /// the archives are dense, scrollable lists.
    enum Scale {
        case call, standard
        var font: Font { self == .call ? .title3 : .body }
    }

    let speaker: DialogueSpeaker
    /// "You", "Future self", or a counterpart's name.
    let name: String
    var scale: Scale = .standard
    /// Playback cursor (Watch) — an accent ring on the line being spoken.
    var isCurrent: Bool = false
    @ViewBuilder var content: () -> Content
    @ViewBuilder var accessory: () -> Accessory

    @AppStorage("futureselfTheme") private var storedTheme = FutureselfTheme.blue.rawValue

    /// Mono is the one palette whose accent is INK rather than a hue, so the
    /// usual 18%-accent wash landed as a barely-there gray: the learner's own
    /// lines stopped being tellable from the fluent self's, which is the only
    /// job this fill has. In mono the user bubble goes solid instead — the
    /// label colour as fill, the background colour as text. Same monochrome
    /// pairing Practice's selected chip uses, and it inverts correctly in dark
    /// mode (white bubble, dark text) rather than sinking into the background.
    private var isMonoUser: Bool {
        speaker.isUser && FutureselfTheme(rawValue: storedTheme) == .mono
    }

    private var bubbleFill: Color {
        if isMonoUser { return Color(.label) }
        return speaker.isUser ? Color.accentColor.opacity(0.18)
                              : Color(.secondarySystemBackground)
    }

    /// Only forced on the solid mono bubble; elsewhere the default cascade
    /// (and any colour a caller set on its own content) stands.
    private var bubbleForeground: Color {
        isMonoUser ? Color(.systemBackground) : Color.primary
    }

    /// The playback ring is accent-coloured — invisible against a solid ink
    /// bubble, so mono draws it in the bubble's own text colour instead.
    private var ringColor: Color {
        isMonoUser ? Color(.systemBackground) : Color.accentColor
    }

    init(speaker: DialogueSpeaker,
         name: String,
         scale: Scale = .standard,
         isCurrent: Bool = false,
         @ViewBuilder content: @escaping () -> Content,
         @ViewBuilder accessory: @escaping () -> Accessory) {
        self.speaker = speaker
        self.name = name
        self.scale = scale
        self.isCurrent = isCurrent
        self.content = content
        self.accessory = accessory
    }

    var body: some View {
        HStack(spacing: 0) {
            // The opposite-side gutter is what actually separates the two
            // speakers' columns — without it both sides span full width and
            // the fills alone read as stripes, not as a back-and-forth.
            if speaker.isUser { Spacer(minLength: 40) }

            VStack(alignment: speaker.alignment, spacing: 4) {
                Text(name)
                    .font(.caption)
                    .foregroundStyle(.secondary)

                content()
                    .font(scale.font)
                    .foregroundStyle(bubbleForeground)
                    // Long lines stay left-ragged even in a trailing bubble —
                    // centre/right-ragged body text is hard to read.
                    .multilineTextAlignment(.leading)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 10)
                    .background(
                        RoundedRectangle(cornerRadius: 14, style: .continuous)
                            .fill(bubbleFill)
                    )
                    .overlay(
                        RoundedRectangle(cornerRadius: 14, style: .continuous)
                            .stroke(isCurrent ? ringColor : .clear, lineWidth: 2)
                    )

                accessory()
            }

            if !speaker.isUser { Spacer(minLength: 40) }
        }
    }
}

extension DialogueLine where Accessory == EmptyView {
    init(speaker: DialogueSpeaker,
         name: String,
         scale: Scale = .standard,
         isCurrent: Bool = false,
         @ViewBuilder content: @escaping () -> Content) {
        self.init(speaker: speaker, name: name, scale: scale, isCurrent: isCurrent,
                  content: content, accessory: { EmptyView() })
    }
}

File: android/app/src/main/java/com/roro/futurevoice/ui/brand/DialogueLine.kt
package com.roro.futurevoice.ui.brand

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

/** Who is speaking. `user` sits trailing on an accent fill; `other` leads. */
enum class DialogueSpeaker { USER, OTHER;
    val isUser: Boolean get() = this == USER
}

/** Text size: the call is read at arm's length, the archives are dense lists. */
enum class DialogueScale { CALL, STANDARD }

/**
 * The ONE dialogue-line surface, shared by every conversation view — Watch's
 * scripted scene, the live call transcript, past-conversation detail
 * (`DialogueLine.swift`).
 *
 * Speaker separation (the name label, which side the line sits on, the bubble
 * fill, the playback ring) is defined HERE and nowhere else, so restyling the
 * conversation UI stays a single-file change. Callers supply their own line
 * content and accessories — never the chrome.
 */
@Composable
fun DialogueLine(
    speaker: DialogueSpeaker,
    name: String,
    modifier: Modifier = Modifier,
    scale: DialogueScale = DialogueScale.STANDARD,
    /** Playback cursor (Watch) — an accent ring on the line being spoken. */
    isCurrent: Boolean = false,
    accessory: @Composable () -> Unit = {},
    content: @Composable () -> Unit,
) {
    val context = LocalContext.current
    val theme = FutureselfTheme.stored(context)
    val accent = theme.tint()
    // Mono is the one palette whose accent is INK rather than a hue, so the
    // usual 18% wash lands as a barely-there gray and the learner's own lines
    // stop being tellable from the fluent self's — the only job this fill
    // has. In mono the user bubble goes solid instead, and it inverts
    // correctly in dark mode rather than sinking into the background.
    val isMonoUser = speaker.isUser && theme == FutureselfTheme.MONO
    val fill = when {
        isMonoUser -> MaterialTheme.colorScheme.onSurface
        speaker.isUser -> accent.copy(alpha = 0.18f)
        else -> MaterialTheme.colorScheme.surfaceVariant
    }
    val foreground = if (isMonoUser) MaterialTheme.colorScheme.surface
    else MaterialTheme.colorScheme.onSurface
    val ring = if (isMonoUser) MaterialTheme.colorScheme.surface else accent

    Row(modifier.fillMaxWidth()) {
        // The opposite-side gutter is what actually separates the two
        // speakers' columns — without it both sides span full width and the
        // fills alone read as stripes, not as a back-and-forth.
        if (speaker.isUser) Spacer(Modifier.width(40.dp))
        Column(
            Modifier.weight(1f),
            horizontalAlignment = if (speaker.isUser) Alignment.End else Alignment.Start,
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text(name, style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            Column(
                Modifier
                    .background(fill, RoundedCornerShape(14.dp))
                    .then(if (isCurrent) Modifier.border(2.dp, ring, RoundedCornerShape(14.dp))
                    else Modifier)
                    .padding(horizontal = 14.dp, vertical = 10.dp)
            ) {
                androidx.compose.runtime.CompositionLocalProvider(
                    androidx.compose.material3.LocalContentColor provides foreground,
                ) {
                    androidx.compose.material3.ProvideTextStyle(
                        // Long lines stay left-ragged even in a trailing
                        // bubble — centre/right-ragged body text is hard to read.
                        (if (scale == DialogueScale.CALL) MaterialTheme.typography.titleMedium
                        else MaterialTheme.typography.bodyLarge).copy(textAlign = TextAlign.Start),
                    ) { content() }
                }
            }
            accessory()
        }
        if (!speaker.isUser) Spacer(Modifier.width(40.dp))
    }
}
